sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/Exchange.scala

```diff
@@ -48,6 +49,10 @@ abstract class Exchange extends UnaryExecNode {
 case class ReusedExchangeExec(override val output: Seq[Attribute], child: Exchange)
   extends LeafExecNode {
 
+  override def outputPartitioning: Partitioning = child.outputPartitioning
```

Comment:
@hvanhovell : continuing our discussion at https://github.com/apache/spark/pull/16985#discussion_r101905768 : 

I found that `ReusedExchangeExec` does not use child's partitioning which was causing the tests to fail. This version of the PR is with your suggestion + works in all cases (all unit tests are passing).